navigation/navigation-compose-material/integration-tests/navigation-demos/src/main/java/androidx/navigation/compose/material/demos/BottomSheetNavDemo.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.navigation.compose.material.demos
+
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.material.Button
+import androidx.compose.material.ExperimentalMaterialApi
+import androidx.compose.material.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.SideEffect
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.navigation.compose.NavHost
+import androidx.navigation.compose.composable
+import androidx.navigation.compose.material.ModalBottomSheetLayout
+import androidx.navigation.compose.material.bottomSheet
+import androidx.navigation.compose.material.rememberBottomSheetNavigator
+import androidx.navigation.compose.rememberNavController
+import androidx.navigation.plusAssign
+import java.util.UUID
+
+private object Destinations {
+    const val Home = "HOME"
+    const val Feed = "FEED"
+    const val Sheet = "SHEET"
+}
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun BottomSheetNavDemo() {
+    val navController = rememberNavController()
+    val bottomSheetNavigator = rememberBottomSheetNavigator(navController = navController)
+    SideEffect { navController.navigatorProvider += bottomSheetNavigator }
```

Comment:
Hello, I hope you don't mind a question from a stranger, but I am working on a similar problem and was referred to your PR as an example.

My question is, is the use of `SideEffect` correct here? Since the block doesn't run until after composition, if you call `NavGraphBuilder.bottomSheet()` immediately, you will get an IllegalStateException. That is to say, I believe the following hypothetical code would fail:

```Kotlin
val navController = rememberNavController()
val bottomSheetNavigator = rememberBottomSheetNavigator(navController = navController)
SideEffect { navController.navigatorProvider += bottomSheetNavigator }
// Note: not wrapped in ModalBottomSheetLayout
NavHost(...) {
  bottomSheet(...) {
    ...
  }
}
```
It would fail with this error:
> java.lang.IllegalStateException: Could not find Navigator with name "BottomSheetNavigator". You must call NavController.addNavigator() for each navigation type.

If `SideEffect` were removed, there would be no error. However I am not sure if it is correct to remove `SideEffect`. Perhaps @ianhanniballake can advise?

I speculate that you haven't seen this error so far, because the ModalBottomSheetLayout begins in "Hidden" state and is thus not part of the composition, thus the `SideEffect` is run before `bottomSheet()`. I suspect the tests pass because they don't use `SideEffect`. But I am speculating; I haven't run the code.

Reply:
`SideEffect` is specifically exactly what should be used, as explained in [b/187443146](https://issuetracker.google.com/187443146) and specifically made possible in #176. The navigation graph is applied as part of a `DisposableEffect`, so this all works perfectly in practice 😄

Reply:
Given the discussion in [b/193555369](https://issuetracker.google.com/issues/193555369), I think I was wrong here and that `SideEffect` is not the right choice. As per the [third comment](https://issuetracker.google.com/issues/193555369#comment3), it is safe to apply the Navigator as part of composition at this point, so let's do that until the better API mentioned in that bug is created.